### ISSUE TITLE
Memory exhausted on Windows10.

### ISSUE THREAD

**Issue by Suhar777777** [ariya/phantomjs#14143]
I opening webpages with casperjs 1.1.0 beta5 from phantomjs 2.1.1. After 87 webpage phantomjs stoped and I saw in command line:" Memory exhausted." In task manager then opened webpages from 1 to 83, memory stable increase from 50 mb to 1050 mb, but then opened 84-87 it down to 900 mb and increase to 1050 mb. And after stoped with: "Memory exhausted". I have 8000mb memory and 11500mb virtual memory. I think I have a lot of memory for casperjs and phantomjs, but I not understend why it is stoped with: "Memory exhausted". Please help!!!

**Comment by mepard** [comment]
Might be a duplicate of #14038. Check out the discussion there.

**Comment by Suhar777777** [comment]
Not helped((( i think not the same. Phatomjs has some limit to use memory. In my way 1000mb. How to increase to use more then 1000mb?

**Comment by Vitallium** [comment]
@Suhar777777 it would be nice to receive a memory dump of PhantomJS from you. Otherwise there are no chances to help you without any sample script or a memory dump.

**Comment by Suhar777777** [comment]
ok

**Comment by Suhar777777** [comment]
but where I can get memory dump of PhantomJS?

**Comment by Vitallium** [comment]
You can use `procdump` to create a dump - https://technet.microsoft.com/en-us/sysinternals/dd996900.aspx
To create a snapshot of PhantomJS: `procdump -ma phantomjs` Remember, you need to execute this command when PhantomJS almost reached the memory limit. Let's say when PhantomJS will consume around 700 Mb, then you must create the dump.

**Comment by Suhar777777** [comment]
Ok. I'll try.

**Comment by Suhar777777** [comment]
0.9gb

**Comment by Suhar777777** [comment]
and what is next step?

**Comment by Vitallium** [comment]
if you created a dump successfully, you need to upload is somewhere :-) Then I'll examine it to find out possible memory leaks, no-disposed resources, etc.

**Comment by Suhar777777** [comment]
Why are you think that it is a memory leaks?

**Comment by Suhar777777** [comment]
I think it is standard situation like a simple web browser. Then you open new pages, and not close it. Memory will be increase. In my way it is the same. But why it not use memory over 1000mb.

**Comment by Vitallium** [comment]
Because I can't reproduce this on my machine. PhantomJS works fine even on 200 pages. It never fails with the mentioned error.

**Comment by Suhar777777** [comment]
I'll try use another browser slimerjs. And if it will be the same, then i linked to memory dump.

**Comment by Suhar777777** [comment]
200 pages? That's what I need!!! I have script to open 200 pages at the same.

**Comment by Suhar777777** [comment]
By the way this 200 pages are dynamical?

**Comment by Suhar777777** [comment]
I used my script with slimerjs and it error was happened, but not the same. [error] [phantom] Wait timeout of 5000ms expired, exiting.Wait timeout of 5000ms expired, exiting. 
And I have not 87 pages, like a PhantomJs, it was 175 pages, but not 200((( Memory was 2000mb. At last pages(130-175) opened very very slow, so that may be error on slimerjs. Speed for open pages on PhantomJs very fast then slimerjs. So use PhantomJs more preferably!

**Comment by Suhar777777** [comment]
that is my script
var WP = require('webpage'), load = 0;
function a() {
    var p = WP.create();
    p.open('http://pt.aliexpress.com/?spm=2114.33020308.1000002.1.QREpI3', function() {
        console.log(load++);
        a();
    });
}
a();

**Comment by Suhar777777** [comment]
Please help! How to fix "Memory exhausted". After 74 page it is happened(((

**Comment by Vitallium** [comment]
You should close Web Pages with `webpage.close` - http://phantomjs.org/api/webpage/method/close.html

Also, I found this is a duplicated issue. See https://github.com/ariya/phantomjs/issues/11390

**Comment by Suhar777777** [comment]
I know that. But I want do without close!

**Comment by Suhar777777** [comment]
Why it happened?

**Comment by Vitallium** [comment]
And why would you expect to have a low (or moderate) memory consumption if you never free it?

**Comment by Suhar777777** [comment]
А по русскому можете написать?

**Comment by Vitallium** [comment]
Написать что? Вы ожидаете, что без освобождения ресурсов все будет нормально? Конечно нет. Ваш код создает новую страницу при каждом вызове `a`, но Вы не удаляете страницу (освобождаете памяти), конечно потребление памяти будет расти. И, в конечном счете, PhantomJS упадет с сообщением о нехватке памяти.

Please use English language in future discussion.
>Only submit a bug report in English language

**Comment by Suhar777777** [comment]
I know all that you writing now. But I not understand why phantomjs use over 1000mb?

**Comment by Suhar777777** [comment]
Sorry not use over 1000mb

**Comment by Suhar777777** [comment]
I have a lot of operative memory 8000mb, then phatomjs stoped with Memory exhausted, I have more then 4000mb free memory

**Comment by Suhar777777** [comment]
May PantomJs has some hiden settings for use many memory 3 - 4 gb?

**Comment by Vitallium** [comment]
No. PhantomJS for Windows is 32bit application. It can't use more than 4Gb RAM.

Please, stop responding to the close thread. This issue tracker primarily serves as the tool to report and track bugs, defects, and development tasks.

**Comment by Suhar777777** [comment]
This is BUGGGGGG for PhantomJS 2.1.1!!!

**Comment by Vitallium** [comment]
The bug is in your code.

**Comment by Suhar777777** [comment]
How can PhantomJs open 200 pages at the same, If it can not open 74 pages in my code!!!

**Comment by Suhar777777** [comment]
Why slimerjs can open 150 pages with my code, and pantomjs not open 75 pages!!! This is BUGGGGGGG fo PhantomJS 2.1.1!!!

**Comment by Vitallium** [comment]
Because people don't open 200 pages at the same time. They use `page.close()` to free the memory and continue working with PhantomJS.

If you need to work with huge amount of opened pages at the same time without closing them - feel free to compile the special build (for example, the 64bit version) for yourself.

**Comment by Suhar777777** [comment]
What have people , I found a problem that does not exist in slimerjs , but you want to focus not notice it !!! So why people buy 10 of gigabytes of RAM , which would not enjoy it???!!! And 1 - 4gb not use pantomjs this is BUGGGGG!!! It use only 0 - 1 gb RAM!!!

**Comment by Vitallium** [comment]
We're sorry. We have an opened task for 64-bit version of PhantomJS for Windows - #11991 . You can keep your eyes on it and get once it is done.

**Comment by Suhar777777** [comment]
I again ask you a simple question. If PhantomJS is a 32 bit application, that needs to use up to 4 GB of RAM , why PhantomJS does not use 4gb RAM fully , and falls with Memory exhausted, when reaching only 1gb of RAM ??? !!!

**Comment by Vitallium** [comment]
How did you measure it? With a task manager? This is not the right tool to use. It doesn't show the real RAM usage.

**Comment by Suhar777777** [comment]
Now I run my script in SlimerJS, and now it is 220 page open at the same!!! And it still RUN!!! Yes task manager I used!

**Comment by zackw** [comment]
1. SlimerJS has a completely different internal architecture.  "SlimerJS uses less memory than PhantomJS when many copies of the same page are opened simultaneously, please improve how PhantomJS handles this" is a valid improvement request, but I suspect one that would have to be addressed within Webkit, at which point my usual Webkit-bug mantra applies: we don't have the manpower to fix Webkit bugs ourselves, please take this up directly with the Webkit developers.

1. 32-bit applications on Windows (even when Windows itself is 64-bit) can, by default, only address 2GB of RAM.  We could try rebuilding it with IMAGE_FILE_LARGE_ADDRESS_AWARE, which would raise the OS-imposed limit to 4GB when run on a 64-bit system, but (a) that might expose other bugs (which is why it's not the default) and (b) since you seem to be hitting a limit at *one* gigabyte, probably this is not the issue.  It is much more likely to be some sort of internal limitation in Webkit.  Again: please take this up directly with the Webkit developers.

**Comment by Suhar777777** [comment]
Finally, sane person. Thanks you! I got it!!!

**Comment by Vitallium** [comment]
@zackw Thank you!. I really appreciate your help. Just a few moments:

1) IMAGE_FILE_LARGE_ADDRESS_AWARE - somehow I thought we already use it. But I just checked it and looks like we don't. We can safely use it, I don't think there are will be side-effects issues with that.
2) WebKit doesn't have an internal limit to 1 GB memory. The problem is that default task manager doesn't show the correct RAM map for a process. I suppose when PhantomJS is about to request additional memory block, it already reached the limit in 2 Gb. But the task manager will still show your that the process may request a memory block.

**Comment by Suhar777777** [comment]
how I can to use IMAGE_FILE_LARGE_ADDRESS_AWARE ?

**Comment by Suhar777777** [comment]
And how correct show RAM. What application do it?

**Comment by Vitallium** [comment]
1) You don't need to use. This is a special instruction to the linker. i.e. this flag must be added on building PhantomJS from sources. 
2) To view the full and complete memory map - use VMMap (https://technet.microsoft.com/en-us/sysinternals/vmmap.aspx?f=255&MSPPError=-2147217396). Launch it and locate the phantomjs process. As an additional work, you can save the memory map and share with us. Thus will help us.

**Comment by mepard** [comment]
I have a website that's getting Memory Exhausted on Windows 10 after around 750 pages. They're consecutive, not simultaneous, and use a single webpage instance (it's with casperjs). I'll see what I can learn about it.

**Comment by Vitallium** [comment]
@mepard it would be nice to get a memory dump of the PhantomJS process to examine it.

**Comment by mepard** [comment]
@Vitallium WIll do.

**Comment by Suhar777777** [comment]
how long time save memory map?

**Comment by zackw** [comment]
@mepard May I suggest you file that as a separate bug so we can discuss it separately from Suhar777777's problem.

**Comment by Vitallium** [comment]
FYI, we have a general issue to track the memory leak problem #11390

**Comment by Suhar777777** [comment]
it is more 5 minutes ago, and it is all saving!

**Comment by Suhar777777** [comment]
max Ram in VMMap identically in task manager not more then 1064mb

**Comment by mahaveer123** [comment]
Using phantomjs on centos(linux) machine. Memory usage is going upto 5 GB.
